man/shutdown.8
'\" t
.TH "SHUTDOWN" "8" "" "systemd 7" "shutdown"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
shutdown \- Halt, power\-off or reboot the machine
.SH "SYNOPSIS"
.HP \w'\fBshutdown\ \fR\fB[OPTIONS...]\fR\fB\ \fR\fB[TIME]\fR\fB\ \fR\fB[WALL...]\fR\ 'u
\fBshutdown \fR\fB[OPTIONS...]\fR\fB \fR\fB[TIME]\fR\fB \fR\fB[WALL...]\fR
.SH "DESCRIPTION"
.PP
\fBshutdown\fR
may be used to halt, power\-off or reboot the machine\&.
.PP
The first argument may be a time string (which is usually
"now")\&. Optionally, this may be followed by a wall message to be sent to all logged\-in users before going down\&.
.PP
The time string may either be in the format
"hh:mm"
for hour/minutes specifying the time to execute the shutdown at, specified in 24h clock format\&. Alternatively it may be in the syntax
"+m"
referring to the specified number of minutes m from now\&.
"now"
is an alias for
"+0", i\&.e\&. for triggering an immediate shutdown\&. If no time argument is specified,
"+1"
is implied\&.
.PP
Note that to specify a wall message you must specify a time argument, too\&.
.PP
If the time argument is used, 5 minutes before the system goes down the
/run/nologin
file is created to ensure that further logins shall not be allowed\&.
.SH "OPTIONS"
.PP
The following options are understood:
.PP
\fB\-\-help\fR
.RS 4
Prints a short help text and exits\&.
.RE
.PP
\fB\-H\fR, \fB\-\-halt\fR
.RS 4
Halt the machine\&.
.RE
.PP
\fB\-P\fR, \fB\-\-poweroff\fR
.RS 4
Power\-off the machine (the default)\&.
.RE
.PP
\fB\-r\fR, \fB\-\-reboot\fR
.RS 4
Reboot the machine\&.
.RE
.PP
\fB\-h\fR
.RS 4
Equivalent to
\fB\-\-poweroff\fR, unless
\fB\-\-halt\fR
is specified\&.
.RE
.PP
\fB\-k\fR
.RS 4
Do not halt, power\-off, reboot, just write wall message\&.
.RE
.PP
\fB\-\-no\-wall\fR
.RS 4
Do not send wall message before halt, power\-off, reboot\&.
.RE
.PP
\fB\-c\fR
.RS 4
Cancel a pending shutdown\&. This may be used cancel the effect of an invocation of
\fBshutdown\fR
with a time argument that is not
"+0"
or
"now"\&.
.RE
.SH "EXIT STATUS"
.PP
On success 0 is returned, a non\-zero failure code otherwise\&.
.SH "SEE ALSO"
.PP
\fBsystemd\fR(1),
\fBsystemctl\fR(1),
\fBhalt\fR(8),
\fBwall\fR(1)


NEWS
uselessd System and Service Manager

Version 8 (2014-11-16)
----------------------
   *  A new feature has been introduced - hard job failures for uselessd
      conditions. By default, if a Condition*= directive in a systemd
      or uselessd service is false, the job is skipped. On the other hand,
      if the file /run/systemd/condition-job-failure exists with a text
      string of Y, T, yes or true, all false conditions will lead to a
      job failure being logged for units on a global level. This is
      similar to systemd-git's (presumably will be in 218) Assert*=
      options, except implemented as a global and volatile flag in the
      /run directory. The default tmpfiles.d configuration creates the
      file with a value of N.

   *  The default tmpfiles.d configuration now respects the legacy
      /forcefsck, /fastboot and /forcequotacheck files and will not
      interfere with them existing.

   *  systemd-sleep has been replaced with a shell script called
      uselessd-sleep. Much simpler, it is currently limited - it
      only executes the /usr/lib/systemd/system-sleep directory once
      and expects shell scripts with an *.sh extension only. In addition,
      it requires an explicitly configured /etc/systemd/sleep.conf file.
      As a result, we provide a stock one. Relevant units have been
      updated to use uselessd-sleep.

   *  systemd-shutdown and systemd-shutdownd will no longer be built if
      --disable-initd is passed to the configure script.

   *  A new configure switch has been introduced: --disable-update-utmp,
      for stripping out the systemd-update-utmp tool from the build.

   *  main-no-init has been significantly simplified. In keeping with
      its purpose as a stage 2 process manager/supervisor, it no longer
      handles things irrelevant to it: /proc/cmdline parsing, crash
      shells, crash VTs, watchdog timers, switch-root handling, daemon
      reexec/keexec/shutdown/reboot/halt/poweroff. In addition, it
      no longer dumps core by default and the --user option has been
      removed, due to main-no-init's intended usage being as a global
      system instance.

      System generator directories are no longer managed and the
      SIGINT/SIGTERM signals are interpreted as a MANAGER_EXIT.

   *  Some updates to systemd-tmpfiles, such as following symlinks
      in lines with w commands and parsing prefixes before suffixes in
      e.g. directories. No major post-208 features as of yet, however.

   *  A few new test cases have been imported from upstream systemd
      with some modifications: test-async, test-set and test-ratelimit.
      More cases for test-util have been added, as well.

   *  Miscellaneous bugfixes and source refactoring.

Version 7 (2014-11-07)
----------------------
   *  A new program has been introduced - uselessd-cgls, which
      is a lightweight alternative to systemd-cgls written in
      Bourne shell. It can be configured out of the build with
      --disable-cgls and comes with a manpage.

   *  Two other new configure switches have been introduced:
      --disable-remount-fs and --disable-sleep, to, respectively
      strip out the systemd-remount-fs tool for rescanning /etc/fstab,
      and the systemd-sleep tool for controlling sleep mode from sysfs
      plus applying user configuration (as well as all respective units
      and manpages).

   *  A wide variety of bugfixes.

Version 6 (2014-10-29)
----------------------
   *  Three new configure options have been added:

      --enable-mudflap for building with GCC Mudflap compiler
      instrumentation on compilers that support it.

      --disable-plymouth for turning off Plymouth event integration
      in units and tty-ask-password-agent.

      --disable-password-agents to strip the systemd-ask-password,
      systemd-reply-password and systemd-tty-ask-password-agent tools
      out of the build. Note that the path units will still be shipped
      if HAVE_SYSV_COMPAT evaluates to true.

   *  The KillMode=mixed unit option for sending SIGTERM to the main
      process of the control group and then a SIGKILL to all the
      remaining ones in the same group, has been backported. The exact
      kill signals are configurable, see man 5 systemd.kill for details.

   *  systemctl has gained an --avoid-bus option which will head
      straight for the /run/systemd/private IPC socket and not do
      any system bus checks, where such a condition is allowed.
      This is a small preliminary step for an intended future
      overhaul of the IPC system.

   *  Some dead code has been removed, shutdownd will no longer
      run an unmount in containers (instead relying on the kernel's
      namespace destructors), main-no-init no longer checks for
      Plymouth running.

Version 5 (2014-10-22)
----------------------
   *  Preliminary experimental support for running a system-wide
      instance of uselessd without being PID1 (init) has been
      introduced.
      See http://uselessd.darknedgy.net/PidNone for details.

   *  Don't allow enabling of masked units.

   *  More vigorous overflow checking in strv_push().

Version 4 (2014-10-06)
----------------------
   *  systemd-analyze, -activate, -cgtop and -delta can now be
      configured out of the build.

   *  systemd-initctl is now only installed when the HAVE_SYSV_COMPAT
      conditional evaluates to true.

   *  The infamous /proc/cmdline "debug" argument fiasco has been
      addressed by renaming it to "systemd.debug", so that it uses
      an exclusive namespace. Currently, upstream systemd resolves
      the issue by keeping the generic term "debug", but simply
      redirecting output to journald as soon as it is available.
      As we do not have it, we implement this.

   *  Various bugfixes from systemd-stable have been backported, like
      basic.target and sysinit.target no longer having a
      RefuseManualStart option.

Version 3 (2014-09-30)
----------------------

src/core/shutdown.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  This file is part of systemd.

  Copyright 2010 ProFUSION embedded systems

  systemd is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation; either version 2.1 of the License, or
  (at your option) any later version.

  systemd is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with systemd; If not, see <http://www.gnu.org/licenses/>.
***/

#include <sys/mman.h>
#include <sys/types.h>
#include <sys/reboot.h>
#include <linux/reboot.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mount.h>
#include <sys/syscall.h>
#include <fcntl.h>
#include <dirent.h>
#include <errno.h>
#include <unistd.h>
#include <signal.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include "missing.h"
#include "log.h"
#include "fileio.h"
#include "umount.h"
#include "util.h"
#include "mkdir.h"
#include "virt.h"
#include "watchdog.h"
#include "killall.h"

#define FINALIZE_ATTEMPTS 50

static int prepare_new_root(void) {
        static const char dirs[] =
                "/run/initramfs/oldroot\0"
                "/run/initramfs/proc\0"
                "/run/initramfs/sys\0"
                "/run/initramfs/dev\0"
                "/run/initramfs/run\0";

        const char *dir;

        if (mount("/run/initramfs", "/run/initramfs", NULL, MS_BIND, NULL) < 0) {
                log_error("Failed to mount bind /run/initramfs on /run/initramfs: %m");
                return -errno;
        }

        if (mount(NULL, "/run/initramfs", NULL, MS_PRIVATE, NULL) < 0) {
                log_error("Failed to make /run/initramfs private mount: %m");
                return -errno;
        }

        NULSTR_FOREACH(dir, dirs)
                if (mkdir_p_label(dir, 0755) < 0 && errno != EEXIST) {
                        log_error("Failed to mkdir %s: %m", dir);
                        return -errno;
                }

        if (mount("/sys", "/run/initramfs/sys", NULL, MS_BIND, NULL) < 0) {
                log_error("Failed to mount bind /sys on /run/initramfs/sys: %m");
                return -errno;
        }

        if (mount("/proc", "/run/initramfs/proc", NULL, MS_BIND, NULL) < 0) {
                log_error("Failed to mount bind /proc on /run/initramfs/proc: %m");
                return -errno;
        }

        if (mount("/dev", "/run/initramfs/dev", NULL, MS_BIND, NULL) < 0) {
                log_error("Failed to mount bind /dev on /run/initramfs/dev: %m");
                return -errno;
        }

        if (mount("/run", "/run/initramfs/run", NULL, MS_BIND, NULL) < 0) {
                log_error("Failed to mount bind /run on /run/initramfs/run: %m");
                return -errno;
        }

        return 0;
}

static int pivot_to_new_root(void) {

        if (chdir("/run/initramfs") < 0) {
                log_error("Failed to change directory to /run/initramfs: %m");
                return -errno;
        }

        /* Work-around for a kernel bug: for some reason the kernel
         * refuses switching root if any file systems are mounted
         * MS_SHARED. Hence remount them MS_PRIVATE here as a
         * work-around.
         *
         * https://bugzilla.redhat.com/show_bug.cgi?id=847418 */
        if (mount(NULL, "/", NULL, MS_REC|MS_PRIVATE, NULL) < 0)
                log_warning("Failed to make \"/\" private mount: %m");

        if (pivot_root(".", "oldroot") < 0) {
                log_error("pivot failed: %m");
                /* only chroot if pivot root succeeded */
                return -errno;
        }

        chroot(".");

        setsid();
        make_console_stdio();

        log_info("Successfully changed into root pivot.");

        return 0;
}

int main(int argc, char *argv[]) {
        _cleanup_free_ char *line = NULL;
        int cmd, r;
        unsigned retries;
        bool need_umount = true, need_swapoff = true, need_loop_detach = true, need_dm_detach = true;
        bool in_container, use_watchdog = false;
        char *arguments[3];

        /* suppress shutdown status output if 'quiet' is used  */
        r = read_one_line_file("/proc/cmdline", &line);
        if (r >= 0) {
                char *w, *state;
                size_t l;

                FOREACH_WORD_QUOTED(w, l, line, state) {
                        if (l == 5 && memcmp(w, "quiet", 5) == 0) {
                                log_set_max_level(LOG_WARNING);
                                break;
                        }
                }
        }

        log_parse_environment();
        log_set_target(LOG_TARGET_CONSOLE); /* syslog will die if not gone yet */
        log_open();

        umask(0022);

        if (getpid() != 1) {
                log_error("Not executed by init (pid 1).");
                r = -EPERM;
                goto error;
        }

        if (argc != 2) {
                log_error("Invalid number of arguments.");
                r = -EINVAL;
                goto error;
        }

        in_container = detect_container(NULL) > 0;

        if (streq(argv[1], "reboot"))
                cmd = RB_AUTOBOOT;
        else if (streq(argv[1], "poweroff"))
                cmd = RB_POWER_OFF;
        else if (streq(argv[1], "halt"))
                cmd = RB_HALT_SYSTEM;
        else if (streq(argv[1], "kexec"))
                cmd = LINUX_REBOOT_CMD_KEXEC;
        else {
                log_error("Unknown action '%s'.", argv[1]);
                r = -EINVAL;
                goto error;
        }

        use_watchdog = !!getenv("WATCHDOG_USEC");

        /* lock us into memory */
        mlockall(MCL_CURRENT|MCL_FUTURE);

        log_info("Sending SIGTERM to remaining processes...");
        broadcast_signal(SIGTERM, true);

        log_info("Sending SIGKILL to remaining processes...");
        broadcast_signal(SIGKILL, true);

        if (in_container) {
                need_swapoff = false;
                need_dm_detach = false;
                need_loop_detach = false;
                need_umount = false;
        }

        /* Unmount all mountpoints, swaps, and loopback devices */
        for (retries = 0; retries < FINALIZE_ATTEMPTS; retries++) {
                bool changed = false;

                if (use_watchdog)
                        watchdog_ping();

                if (need_umount) {
                        log_info("Unmounting file systems.");
                        r = umount_all(&changed);
                        if (r == 0) {
                                need_umount = false;
                                log_info("All filesystems unmounted.");
                        } else if (r > 0)
                                log_info("Not all file systems unmounted, %d left.", r);
                        else
                                log_error("Failed to unmount file systems: %s", strerror(-r));
                }

                if (need_swapoff) {
                        log_info("Deactivating swaps.");
                        r = swapoff_all(&changed);
                        if (r == 0) {
                                need_swapoff = false;
                                log_info("All swaps deactivated.");
                        } else if (r > 0)
                                log_info("Not all swaps deactivated, %d left.", r);
                        else
                                log_error("Failed to deactivate swaps: %s", strerror(-r));
                }

                if (need_loop_detach) {
                        log_info("Detaching loop devices.");
                        r = loopback_detach_all();
                        if (r == 0) {
                                need_loop_detach = false;
                                log_info("All loop devices detached.");
                        } else if (r > 0)
                                log_info("Not all loop devices detached, %d left.", r);
                        else
                                log_error("Failed to detach loop devices: %s", strerror(-r));
                }

                if (need_dm_detach) {
                        log_info("Detaching DM devices.");
                        r = dm_detach_all();
                        if (r == 0) {
                                need_dm_detach = false;
                                log_info("All DM devices detached.");
                        } else if (r > 0)
                                log_info("Not all DM devices detached, %d left.", r);
                        else
                                log_error("Failed to detach DM devices: %s", strerror(-r));
                }

                if (!need_umount && !need_swapoff && !need_loop_detach && !need_dm_detach) {
                        if (retries > 0)
                                log_info("All filesystems, swaps, loop devices, DM devices detached.");
                        /* Yay, done */
                        break;
                }

                /* If in this iteration we didn't manage to
                 * unmount/deactivate anything, we simply give up */
                if (!changed) {
                        log_error("Cannot finalize remaining file systems and devices, giving up.");
                        break;
                }

                log_debug("Couldn't finalize remaining file systems and devices after %u retries, trying again.", retries+1);
        }

        if (retries >= FINALIZE_ATTEMPTS)
                log_error("Too many iterations, giving up.");
        else
                log_info("Storage is finalized.");

        arguments[0] = NULL;
        arguments[1] = argv[1];
        arguments[2] = NULL;
        execute_directory(SYSTEM_SHUTDOWN_PATH, NULL, arguments);

        if (!in_container && !in_initrd() &&
            access("/run/initramfs/shutdown", X_OK) == 0) {

                if (prepare_new_root() >= 0 &&
                    pivot_to_new_root() >= 0) {

                        log_info("Returning to initrd...");

                        execv("/shutdown", argv);
                        log_error("Failed to execute shutdown binary: %m");
                }
        }

        /* The kernel will automaticall flush ATA disks and suchlike
         * on reboot(), but the file systems need to be synce'd
         * explicitly in advance. So let's do this here, but not
         * needlessly slow down containers. */
        if (!in_container)
                sync();

        if (cmd == LINUX_REBOOT_CMD_KEXEC) {

                if (!in_container) {
                        /* We cheat and exec kexec to avoid doing all its work */
                        pid_t pid = fork();

                        if (pid < 0)
                                log_error("Could not fork: %m. Falling back to normal reboot.");
                        else if (pid > 0) {
                                wait_for_terminate_and_warn("kexec", pid);
                                log_warning("kexec failed. Falling back to normal reboot.");
                        } else {
                                /* Child */
                                const char *args[3] = { KEXEC, "-e", NULL };
                                execv(args[0], (char * const *) args);
                                return EXIT_FAILURE;
                        }
                }

                cmd = RB_AUTOBOOT;
        }

        reboot(cmd);

        if (errno == EPERM && in_container) {
                /* If we are in a container, and we lacked
                 * CAP_SYS_BOOT just exit, this will kill our
                 * container for good. */
                log_error("Exiting container.");
                exit(0);
        }

        log_error("Failed to invoke reboot(): %m");
        r = -errno;

  error:
        log_error("Critical error while doing system shutdown: %s", strerror(-r));

        freeze();
        return EXIT_FAILURE;
}
